Kubecost Integrations - Postgres

thomasvn opened the pull request:
## What does this PR change?

Introduces Kubecost Integrations - Postgres. Configurable via k8s secret and k8s configmap.

## Does this PR rely on any other PRs?

* Helm change for https://github.com/kubecost/kubecost-cost-model/pull/2327
* Also this https://github.com/kubecost/kubecost-cost-model/pull/2386
* ~Helm change for https://github.com/kubecost/kubecost-cost-model/pull/2308 and https://github.com/kubecost/kubecost-cost-model/pull/2317~
* ~Supersedes https://github.com/kubecost/cost-analyzer-helm-chart/pull/3171. Previously the Postgres Integration created a k8s cronjob. Now, we run the cronjob directly inside the `aggregator` container.~

## How does this PR impact users? (This is the kind of thing that goes in release notes!)

Introduces Kubecost Integrations - Postgres

## Links to Issues or tickets this PR addresses or fixes

Implements https://team.atlassian.com/project/KUBEC-6

## What risks are associated with merging this PR? What is required to fully test this PR?

N/A. This is a new feature and we've tested a few edge cases on how the backend handles misconfigured parameters.

## How was this PR tested?

```yaml
global:
  integrations:
    postgres:
      enabled: true
      runInterval: "1m"
      databaseHost: "REDACTED.postgres.database.azure.com"
      databasePort: "5432"
      databaseName: "postgres"
      databaseUser: "REDACTED"
      databasePassword: "REDACTED"
      queryConfigs:
        allocations:
          - databaseTable: "kubecost_allocation_data"
            window: "7d"
            aggregate: "namespace"
            idle: true
            shareIdle: false
            shareNamespaces: ""
            shareLabels: ""
          - databaseTable: "kubecost_allocation_data2"
            window: "10d"
            aggregate: "controller"
            idle: true
            shareIdle: false
            shareNamespaces: ""
            shareLabels: ""
          - databaseTable: "kubecost_allocation_data3"
            window: "30d"
            aggregate: "cluster"
            idle: true
            shareIdle: false
            shareNamespaces: ""
            shareLabels: ""
        assets:
          - databaseTable: "kubecost_assets_data"
            window: "7d"
            aggregate: "cluster"
        cloudCosts:
          - databaseTable: "kubecost_cloudcosts_data"
            window: "7d"
            aggregate: "service"
kubecostAggregator:
  deployMethod: "statefulset"
kubecostModel:
  federatedStorageConfigSecret: "federated-store"
kubecostProductConfigs:
  productKey:
    enabled: true
    key: "my-enterprise-key-here"
prometheus:
  server:
    global:
      external_labels:
        cluster_id: "kubecost-thomasn-integrations"
```

```sh
helm upgrade -i thomasn-kc-int ~/kubecost/cost-analyzer-helm-chart/cost-analyzer -f values.yaml
```

When installing into this namespace, validated that the secret mounted to `/var/configs/integrations/postgres-creds/creds.json` and configmap mounted to `/var/configs/integrations/postgres-queries/kubecost-queries.json`.

## Have you made an update to documentation? If so, please provide the corresponding PR.

https://github.com/kubecost/docs/pull/1006

thomasvn committed:
Init

chipzoller commented:
Draft. This has hard upstream dependencies.

thomasvn committed:
Restructure values.yaml configurations to be configurable for
allocations queries, assets queries, and cloudcosts queries.

thomasvn committed:
Fix env vars for Postgres Db

thomasvn committed:
Move environment variables to aggregator.

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres

thomasvn committed:
Fix lint error.

thomasvn committed:
Merge branch 'thomasn/kubecost-integrations-postgres' of github.com:kubecost/cost-analyzer-helm-chart into thomasn/kubecost-integrations-postgres

thomasvn commented:
One thing that could be nice to add to this PR, is to give the user an option to specify a k8s secret holding the Postgres DB password. Or optionally pass provide the password via Helm values and we template a secret for them.

Not sure what the priority of this ^^ is, because by the next release I hope to move all db configs to k8s secrets and all query configs to k8s configmaps as opposed to environment variables.

chipzoller commented:
I would suggest we not worry about that quite yet to get this out the door, but agree it will probably be requested. For the next iteration, we should consume it via a volume though and not an env var as consuming secrets via env vars is generally not a recommended practice since they can be echoed out in log statements.

michaelmdresser commented on cost-analyzer/templates/_helpers.tpl:
```diff
@@ -1006,6 +1006,44 @@ Begin Kubecost 2.0 templates
+      value: {{ quote .Values.global.integrations.postgres.databaseName }}
+    - name: INT_PG_DB_USER
+      value: {{ quote .Values.global.integrations.postgres.databaseUser }}
+    - name: INT_PG_DB_PASS
+      value: {{ quote .Values.global.integrations.postgres.databasePassword }}
```
I know you're aware of this, but it still makes me uncomfortable that a plaintext password would end up in someone's Helm values. Should we be shipping a feature that encourages a bad security practice like that? And more practically, will anyone adopt the feature in this state?

thomasvn replied:
Definitely agree that this isn't the right long term approach. I had prototyped a minor improvement by doing:

```yaml
    - name: INT_PG_DB_PASS
      valueFrom:
        secretKeyRef:
          name: pg-pass
          key: password
```

however, even this wouldn't be the fully correct solution (for reasons Chip mentioned above). I'm a bit wary to deliver this half baked solution, then have to walk it back and replace it with the fully correct solution. I think that we can get a fully correct solution to build nicely on top of these configs and be made available in a patch release or v2.3? Thoughts?

@chipzoller Let me know what you think too! If we're all ok with this, I'd love to merge and test. In the meantime, I'll be working on the full solution here.

chipzoller replied:
Unless there is, for some reason, an absolute hard requirement this _must_ be in 2.2, I would prefer we hold until 2.3 where we get the security aspect a little better so we avoid breaking changes right out of the gate. And as we have already passed the agreed upon go/no-go time, I think it's the better thing to do. Releasing this in a patch to 2.2 would not really be appropriate as this is a definite feature.

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres

thomasvn committed:
Update to use configmap/secret

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres

thomasvn commented:
@chipzoller Could I get a review on this PR? This introduces Kubecost Integrations and allows configurability via Configmap and K8s secret.

Future work includes:
1. The ability for a user to provide their own Secret or Configmap (instead of the default templated one)
2. The ability to configure query parameters for Assets and CloudCosts.

thomasvn committed:
Include k8s secret and configmap.

chipzoller commented on cost-analyzer/values.yaml:
```diff
@@ -249,6 +249,34 @@ global:
+      allocations:
+        enabled: true
+        queryConfigs:
+          - databaseTable: "kubecost_allocation_data"
+            window: "7d"
+            aggregate: "namespace"
+            idle: "true"
+            shareIdle: "false"
+            shareNamespaces: ""
+            shareLabels: ""
```
Main question here is what is the behavior inside cost-model if the inputs provided here are wrong, for example:

1. the value of a param is of the wrong type
2. a param is omitted
3. the API changes to include a new field (required) and it is omitted (see 2)
4. a key in a param here is wrong (ex., "windw" instead of "window")

Just making sure that we're somehow account for these, logging something helpful and, above all, not crashing/panicing. I realize this isn't exactly a question for this Helm change, but it's obviously related.

thomasvn replied:
As an update, the Postgres Integration now runs inside the Aggregator container. It runs as a separate goroutine so if it were to crash all other aggregator processes would carry on as usual!

If (1) (2) or (3) occurs, the API request to the aggregator would fail and the job would exit with an error message of the failure.

If (4) occurs, it would be similar to omitting a parameter and and the job would exit if the API request fails without that parameter.

chipzoller commented on cost-analyzer/templates/integrations-postgres-queries-configmap.yaml:
```diff
@@ -0,0 +1,18 @@
+        {{- if .Values.global.integrations.postgres.allocations.enabled }}
+        "allocations": {{- .Values.global.integrations.postgres.allocations.queryConfigs | toJson | nindent 10 }}
+        {{- else }}
+        "allocations": []
```
Would we really have to include an empty array if allocations isn't enabled? Or can we just omit entirely?

thomasvn replied:
Let me double check how the backend handles cases where the "allocations" key does not exist in `/var/configs/integrations/postgres-queries/kubecost-queries.json`

thomasvn replied:
I've tested this in the backend, and omitting entirely is ok. I've made the corresponding Helm change.

chipzoller commented on cost-analyzer/templates/aggregator-statefulset.yaml:
So this will only work in a configuration in which aggregator runs as a SS?

thomasvn replied:
Yes that's the current intended behavior. My thinking was that this is an Enterprise-only feature, and most Enterprise-scale users are recommended to and are currently (based on my observation) running the SS.

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres

thomasvn committed:
Cleanup configmap.

thomasvn committed:
Update structure of `values.yaml` and how it gets templated to kubecost-integrations-postgres-queries

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres

thomasvn commented:
@chipzoller I've updated the `values.yaml` to reflect what we discussed on Slack. The secret & configmap are also updated accordingly. When you get a chance, would love a final review on this so we can merge it!

To address in a followup PR. We should allow users to specify their own k8s secrets which they have created containing the Postgres creds.

thomasvn committed:
Add ability to specify your own k8s secret.

thomasvn committed:
Merge branch 'thomasn/kubecost-integrations-postgres' of github.com:kubecost/cost-analyzer-helm-chart into thomasn/kubecost-integrations-postgres

thomasvn commented:
@chipzoller I've just added the ability for users to configure their own k8s secret via `.Values.global.integrations.postgres.databaseSecretName`.

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres

chipzoller reviewed (approved):
Nice job here, @thomasvn! Looks good to me. Anxious to play with this. Let's start testing next week.

thomasvn committed:
Merge branch 'develop' into thomasn/kubecost-integrations-postgres
